File: frontend/src/app/services/auth/auth.service.ts
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { AngularFireAuth } from '@angular/fire/auth';
import firebase from 'firebase/app';
import { Router } from '@angular/router';
import { skip, take } from 'rxjs/operators';
import { AuthUser } from '../../models/auth/auth-user';
import { LoginData } from '../../models/auth/login-data';
import { HttpInternalService } from '../http-internal.service';
import { RegisterData } from '../../models/auth/register-data';

type AuthState = AuthUser | null;

interface RegisterWithEmailParams {
  registerData: RegisterData;
  redirect?: string;
}

interface SignInWithEmailParams {
  email: string;
  password: string;
  remember?: boolean;
  redirect?: string;
}

interface SignInWithSocialsParams {
  remember?: boolean;
  redirect?: string;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private _currentAuthState: AuthState = null;
  private _authStateSubject: ReplaySubject<AuthState> = new ReplaySubject<AuthState>(1);

  constructor(
    private _httpService: HttpInternalService,
    private _fireAuth: AngularFireAuth,
    private _router: Router
  ) {
    this._fireAuth.authState.subscribe(
      (firebaseUser) => this._updateAuthState(firebaseUser)
    );
  }

  get authenticated() {
    return this._currentAuthState !== null;
  }

  getAuthStateObservable() {
    return this._authStateSubject.asObservable();
  }

  getCurrentTokenObservable() {
    return this._fireAuth.idToken;
  }

  async registerWithEmail({ registerData, redirect }: RegisterWithEmailParams) {
    const response = await this._httpService.postFullRequest(
      '/api/auth/register', registerData
    ).toPromise();

    if (!response.ok) {
      return;
    }

    await this.signInWithEmail({
      email: registerData.email,
      password: registerData.password
    });

    if (redirect !== undefined) {
      this._redirectAfterAuthUpdate(redirect);
    }
  }

  async signInWithEmail({
    email,
    password,
    remember,
    redirect
  }: SignInWithEmailParams) {
    await this._setPersistence(remember ?? true);
    await this._fireAuth.signInWithEmailAndPassword(email, password);

    if (redirect !== undefined) {
      this._redirectAfterAuthUpdate(redirect);
    }
  }

  async signInWithGoogle({ remember, redirect }: SignInWithSocialsParams) {
    await this._setPersistence(remember ?? true);

    const provider = new firebase.auth.GoogleAuthProvider();
    provider.addScope('email');
    provider.addScope('profile');
    await this._fireAuth.signInWithPopup(provider);

    if (redirect !== undefined) {
      this._redirectAfterAuthUpdate(redirect);
    }
  }

  signInWithApple = () => {
    // TODO Implement
  };

  signInWithFacebook = () => {
    // TODO Implement
  };

  signOut() {
    return this._fireAuth.signOut()
      .then(() => this._router.navigateByUrl('/login'));
  }

  private async _updateAuthState(firebaseUser: firebase.User | null) {
    if (!firebaseUser) {
      this._setAuthState(null);
      return;
    }

    if (!AuthService.validateFirebaseUser(firebaseUser)) {
      await this.signOut();
      return;
    }

    const loginData: LoginData = {
      accessToken: await firebaseUser.getIdToken(),
      firebaseId: firebaseUser.uid,
      userName: firebaseUser.displayName!,
      email: firebaseUser.email!
    };

    const response = await this._httpService.postFullRequest(
      '/api/auth/login', loginData
    ).toPromise();

    if (!response.ok) {
      await this.signOut();
      return;
    }

    const tokenResult = await firebaseUser.getIdTokenResult(true);

    const authUser: AuthUser = {
      id: tokenResult.claims.id,
      firebaseId: firebaseUser.uid,
      role: tokenResult.claims.role,
      userName: firebaseUser.displayName!,
      email: firebaseUser.email!,
      accessToken: tokenResult.token,
      refreshToken: firebaseUser.refreshToken
    };

    this._setAuthState(authUser);
  }

  static validateFirebaseUser(firebaseUser: firebase.User) {
    return firebaseUser.email !== undefined && firebaseUser.displayName !== undefined;
  }

  private _setAuthState(authState: AuthState) {
    this._currentAuthState = authState;
    this._authStateSubject.next(authState);
  }

  private _setPersistence(value: boolean) {
    return this._fireAuth.setPersistence(value ? 'local' : 'session');
  }

  private _redirectAfterAuthUpdate(url: string) {
    return this.getAuthStateObservable()
      .pipe(
        skip(1), // skip current authState
        take(1)
      )
      .subscribe(
        (authState) => {
          if (authState !== null) {
            this._router.navigateByUrl(url);
          }
        }
      );
  }
}

File: mobile/lib/services/auth/auth_service.dart
import 'dart:async';
import 'dart:io';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:get_it/get_it.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:injectable/injectable.dart';
import 'package:perflow/models/auth/auth_user.dart';
import 'package:perflow/models/auth/login_data.dart';
import 'package:perflow/models/auth/register_data.dart';
import 'package:perflow/models/enums/user_role.dart';
import 'package:perflow/services/auth/auth_api.dart';
import 'package:rxdart/rxdart.dart';

@Singleton(signalsReady: true)
class AuthService {
  final AuthApi _authApi;
  late final FirebaseAuth _auth;

  AuthUser? _currentAuthState;
  final _authStateSubject = ReplaySubject<AuthUser?>(maxSize: 1);
  
  AuthService(
    this._authApi
  ) {
    _initialize();
  }

  Future<void> _initialize() async {
    final app = await Firebase.initializeApp();
    _auth = FirebaseAuth.instanceFor(app: app);

    _auth.authStateChanges().listen(_handleAuthStateChange);

    GetIt.instance.signalReady(this);
  }

  bool get isAuthenticated => _currentAuthState != null;

  AuthUser? get currentAuthState => _currentAuthState;

  Stream<AuthUser?> getAuthStateStream() {
    return _authStateSubject.stream;
  }

  Future<String?> getToken() async {
    if(!isAuthenticated) {
      return null;
    }

    return await _auth.currentUser?.getIdToken();
  }

  Future<void> registerWithEmail(RegisterData registerData) async {
    final registerResponse = await _authApi.register(registerData);

    if(registerResponse.statusCode != HttpStatus.ok) {
      return;
    }

    await signInWithEmail(
      email: registerData.email,
      password: registerData.password
    );
  }

  Future<void> signInWithEmail({
    required String email,
    required String password
  }) {
    return _auth.signInWithEmailAndPassword(email: email, password: password);
  }

  Future<void> signInWithGoogle() async {
    try {
      final googleUser = await GoogleSignIn(
          scopes: ['email', 'profile']
      ).signIn();

      if(googleUser == null) {
        return signOut();
      }

      final googleAuth = await googleUser.authentication;

      final credential = GoogleAuthProvider.credential(
          accessToken: googleAuth.accessToken,
          idToken: googleAuth.idToken
      );

      await _auth.signInWithCredential(credential);
    }
    catch (e) {
      signOut();
      rethrow;
    }
  }

  Future<void> signOut() {
    return _auth.signOut();
  }

  Future<void> _handleAuthStateChange(User? firebaseUser) async {
    if(firebaseUser == null) {
      _setAuthState(null);
      return;
    }

    if(firebaseUser.email == null || firebaseUser.displayName == null) {
      await signOut();
      return;
    }

    final loginData = LoginData(
      accessToken: await firebaseUser.getIdToken(),
      firebaseId: firebaseUser.uid,
      userName: firebaseUser.displayName!,
      email: firebaseUser.email!
    );

    try {
      final loginResponse = await _authApi.login(loginData);

      if(loginResponse.statusCode != HttpStatus.ok) {
        await signOut();
        return;
      }

      final tokenResult = await firebaseUser.getIdTokenResult(true);

      final authUser =  AuthUser(
          id: tokenResult.claims!['id'],
          firebaseId: firebaseUser.uid,
          role: UserRole.values[tokenResult.claims!['role']],
          userName: firebaseUser.displayName!,
          email: firebaseUser.email!,
          accessToken: tokenResult.token!,
          refreshToken: firebaseUser.refreshToken!
      );

      _setAuthState(authUser);
    }
    catch (_) {
      signOut();
    }
  }

  void _setAuthState(AuthUser? state) {
    _currentAuthState = state;
    _authStateSubject.add(state);
  }
}
